app/grandchallenge/reader_studies/forms.py

```diff
@@ -478,41 +480,174 @@ def clean_ground_truth(self):
+                self.fields[slug] = self._get_json_field(
+                    interface, current_value
+                )
+
+        order = (
+            instance.order if instance else reader_study.next_display_set_order
+        )
+        self.fields["order"] = IntegerField(initial=order)
+
+    def _get_default_field(self, interface, current_value):
+        return InterfaceFormField(
+            instance=interface,
+            initial=current_value.value if current_value else None,
+            required=False,
+            user=self.user,
+        ).field
+
+    def _get_json_field(self, interface, current_value):
+        return self._get_default_field(interface, current_value)
+
+    def _get_image_field(self, interface, values, current_value):
+        return ModelChoiceField(
+            queryset=ComponentInterfaceValue.objects.filter(id__in=values),
+            initial=current_value,
+            required=False,
+            widget=SelectUploadWidget(
+                attrs={
+                    "reader_study_slug": self.reader_study.slug,
+                    "display_set_pk": self.instance.pk,
+                    "interface_pk": interface.pk,
+                }
+            ),
+        )
+
+    def _get_file_field(self, interface, values, current_value):
+        return self._get_image_field(interface, values, current_value)
+
+
+class DisplaySetUpdateForm(DisplaySetBaseForm):
+    def __init__(self, *args, instance, user, **kwargs):
+        super().__init__(
+            *args,
+            instance=instance,
+            reader_study=instance.reader_study,
+            user=user,
+            **kwargs,
+        )
+
+
+class DisplaySetCreateForm(DisplaySetBaseForm):
+    def __init__(self, *args, instance=None, reader_study, user, **kwargs):
+        super().__init__(
+            *args,
+            instance=None,
+            reader_study=reader_study,
+            user=user,
+            **kwargs,
+        )
+
+    def _get_image_field(self, interface, values, current_value):
+        return self._get_default_field(interface, current_value)
+
+
+class FileForm(Form):
+
+    user_upload = ModelChoiceField(
+        label="File",
+        queryset=None,
+    )
+
+    def __init__(
+        self, *args, user, display_set, interface, instance=None, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.fields["user_upload"].widget = UserUploadSingleWidget(
+            allowed_file_types=interface.file_mimetypes
+        )
+        self.fields["user_upload"].queryset = get_objects_for_user(
+            user, "uploads.change_userupload", accept_global_perms=False
+        ).filter(status=UserUpload.StatusChoices.COMPLETED)
+        self.interface = interface
+        self.display_set = display_set
+
+    def save(self):
+        civ = self.display_set.values.get(interface=self.interface)
+        user_upload = self.cleaned_data["user_upload"]
+        user_upload.copy_object(to_field=civ.file)
```

Comment:
This cannot be done in the request-response cycle. Needs to be done in an async celery task.

Reply:
The only thing I'm unsure of here is how to let the user know if their file does not pass the civ validation, for instance if it does not match the schema provided for the interface. We cannot update the upload session status, as this is merely a UserUpload instance, which get deleted.

Reply:
A notification?